File: torchvision/models/googlenet.py
import torch
import torch.nn as nn
import torch.nn.functional as F
from torch.utils import model_zoo

__all__ = ['GoogLeNet', 'googlenet']

model_urls = {
    # GoogLeNet ported from TensorFlow
    'googlenet': 'https://download.pytorch.org/models/googlenet-1378be20.pth',
}


def googlenet(pretrained=False, **kwargs):
    r"""GoogLeNet (Inception v1) model architecture from
    `"Going Deeper with Convolutions" <http://arxiv.org/abs/1409.4842>`_.
    Args:
        pretrained (bool): If True, returns a model pre-trained on ImageNet
    """
    if pretrained:
        if 'transform_input' not in kwargs:
            kwargs['transform_input'] = True
        kwargs['init_weights'] = False
        model = GoogLeNet(**kwargs)
        model.load_state_dict(model_zoo.load_url(model_urls['googlenet']))
        return model

    return GoogLeNet(**kwargs)


class GoogLeNet(nn.Module):

    def __init__(self, num_classes=1000, aux_logits=True, transform_input=False, init_weights=True):
        super(GoogLeNet, self).__init__()
        self.aux_logits = aux_logits
        self.transform_input = transform_input

        self.conv1 = BasicConv2d(3, 64, kernel_size=7, stride=2, padding=3)
        self.maxpool1 = nn.MaxPool2d(3, stride=2, ceil_mode=True)
        self.conv2 = BasicConv2d(64, 64, kernel_size=1)
        self.conv3 = BasicConv2d(64, 192, kernel_size=3, padding=1)
        self.maxpool2 = nn.MaxPool2d(3, stride=2, ceil_mode=True)

        self.inception3a = Inception(192, 64, 96, 128, 16, 32, 32)
        self.inception3b = Inception(256, 128, 128, 192, 32, 96, 64)
        self.maxpool3 = nn.MaxPool2d(3, stride=2, ceil_mode=True)

        self.inception4a = Inception(480, 192, 96, 208, 16, 48, 64)
        self.inception4b = Inception(512, 160, 112, 224, 24, 64, 64)
        self.inception4c = Inception(512, 128, 128, 256, 24, 64, 64)
        self.inception4d = Inception(512, 112, 144, 288, 32, 64, 64)
        self.inception4e = Inception(528, 256, 160, 320, 32, 128, 128)
        self.maxpool4 = nn.MaxPool2d(2, stride=2, ceil_mode=True)

        self.inception5a = Inception(832, 256, 160, 320, 32, 128, 128)
        self.inception5b = Inception(832, 384, 192, 384, 48, 128, 128)
        if aux_logits:
            self.aux1 = InceptionAux(512, num_classes)
            self.aux2 = InceptionAux(528, num_classes)
        self.avgpool = nn.AdaptiveAvgPool2d((1, 1))
        self.dropout = nn.Dropout(0.4)
        self.fc = nn.Linear(1024, num_classes)

        if init_weights:
            self._initialize_weights()

    def _initialize_weights(self):
        for m in self.modules():
            if isinstance(m, nn.Conv2d):
                nn.init.xavier_uniform_(m.weight)
                if m.bias is not None:
                    nn.init.constant_(m.bias, 0.2)
            elif isinstance(m, nn.Linear):
                nn.init.xavier_uniform_(m.weight)
                nn.init.constant_(m.bias, 0)
            elif isinstance(m, nn.BatchNorm2d):
                nn.init.constant_(m.weight, 1)
                nn.init.constant_(m.bias, 0)

    def forward(self, x):
        if self.transform_input:
            x_ch0 = torch.unsqueeze(x[:, 0], 1) * (0.229 / 0.5) + (0.485 - 0.5) / 0.5
            x_ch1 = torch.unsqueeze(x[:, 1], 1) * (0.224 / 0.5) + (0.456 - 0.5) / 0.5
            x_ch2 = torch.unsqueeze(x[:, 2], 1) * (0.225 / 0.5) + (0.406 - 0.5) / 0.5
            x = torch.cat((x_ch0, x_ch1, x_ch2), 1)

        # N X 3 X 224 X 224
        x = self.conv1(x)
        # N X 64 X 112 X 112
        x = self.maxpool1(x)
        # N X 64 X 56 X 56
        x = self.conv2(x)
        # N X 64 X 56 X 56
        x = self.conv3(x)
        # N X 192 X 56 X 56
        x = self.maxpool2(x)

        # N X 192 X 28 X 28
        x = self.inception3a(x)
        # N X 256 X 28 X 28
        x = self.inception3b(x)
        # N X 480 X 28 X 28
        x = self.maxpool3(x)
        # N X 480 X 14 X 14
        x = self.inception4a(x)
        # N X 512 x 14 X 14
        if self.training and self.aux_logits:
            aux1 = self.aux1(x)

        x = self.inception4b(x)
        # N X 512 X 14 X 14
        x = self.inception4c(x)
        # N X 512 X 14 X 14
        x = self.inception4d(x)
        # N X 528 X 14 X 14
        if self.training and self.aux_logits:
            aux2 = self.aux2(x)

        x = self.inception4e(x)
        # N X 832 X 14 X 14
        x = self.maxpool4(x)
        # N X 832 X 7 X 7
        x = self.inception5a(x)
        # N X 832 X 7 X 7
        x = self.inception5b(x)
        # N X 1024 X 7 X 7

        x = self.avgpool(x)
        # N X 1024 X 1 X 1
        x = x.view(x.size(0), -1)
        # N X 1024
        x = self.dropout(x)
        # N X 1024
        x = self.fc(x)
        # N X 1000 (num_classes)
        if self.training and self.aux_logits:
            return aux1, aux2, x
        return x


class Inception(nn.Module):

    def __init__(self, in_channels, ch1x1, ch3x3red, ch3x3, ch5x5red, ch5x5, pool_proj):
        super(Inception, self).__init__()

        self.branch1 = BasicConv2d(in_channels, ch1x1, kernel_size=1)

        self.branch2 = nn.Sequential(
            BasicConv2d(in_channels, ch3x3red, kernel_size=1),
            BasicConv2d(ch3x3red, ch3x3, kernel_size=3, padding=1)
        )

        self.branch3 = nn.Sequential(
            BasicConv2d(in_channels, ch5x5red, kernel_size=1),
            BasicConv2d(ch5x5red, ch5x5, kernel_size=3, padding=1)
        )

        self.branch4 = nn.Sequential(
            nn.MaxPool2d(kernel_size=3, stride=1, padding=1, ceil_mode=True),
            BasicConv2d(in_channels, pool_proj, kernel_size=1)
        )

    def forward(self, x):
        branch1 = self.branch1(x)
        branch2 = self.branch2(x)
        branch3 = self.branch3(x)
        branch4 = self.branch4(x)

        outputs = [branch1, branch2, branch3, branch4]
        return torch.cat(outputs, 1)


class InceptionAux(nn.Module):

    def __init__(self, in_channels, num_classes):
        super(InceptionAux, self).__init__()
        self.conv = BasicConv2d(in_channels, 128, kernel_size=1)

        self.fc1 = nn.Linear(2048, 1024)
        self.fc2 = nn.Linear(1024, num_classes)

    def forward(self, x):
        # aux1: N X 512 X 14 X 14, aux2: N X 528 X 14 X 14
        x = F.adaptive_avg_pool2d(x, (4, 4))
        # aux1: N X 512 X 4 X 4, aux2: N X 528 X 4 X 4
        x = self.conv(x)
        # N X 128 X 4 X 4
        x = x.view(x.size(0), -1)
        # N X 2048
        x = F.relu(self.fc1(x), inplace=True)
        # N X 2048
        x = F.dropout(x, 0.7, training=self.training)
        # N X 2048
        x = self.fc2(x)
        # N X 1024

        return x


class BasicConv2d(nn.Module):

    def __init__(self, in_channels, out_channels, **kwargs):
        super(BasicConv2d, self).__init__()
        self.conv = nn.Conv2d(in_channels, out_channels, bias=False, **kwargs)
        self.bn = nn.BatchNorm2d(out_channels, eps=0.001)

    def forward(self, x):
        x = self.conv(x)
        x = self.bn(x)
        return F.relu(x, inplace=True)

File: torchvision/csrc/models/googlenet.cpp
#include "googlenet.h"

#include "modelsimpl.h"

namespace vision {
namespace models {

using Options = torch::nn::Conv2dOptions;

namespace _googlenetimpl {
BasicConv2dImpl::BasicConv2dImpl(torch::nn::Conv2dOptions options) {
  options.with_bias(false);
  conv = torch::nn::Conv2d(options);
  bn = torch::nn::BatchNorm(
      torch::nn::BatchNormOptions(options.output_channels()).eps(0.001));

  register_module("conv", conv);
  register_module("bn", bn);
}

torch::Tensor BasicConv2dImpl::forward(torch::Tensor x) {
  x = conv->forward(x);
  x = bn->forward(x);
  return x.relu_();
}

InceptionImpl::InceptionImpl(
    int64_t in_channels,
    int64_t ch1x1,
    int64_t ch3x3red,
    int64_t ch3x3,
    int64_t ch5x5red,
    int64_t ch5x5,
    int64_t pool_proj) {
  branch1 = BasicConv2d(Options(in_channels, ch1x1, 1));

  branch2->push_back(BasicConv2d(Options(in_channels, ch3x3red, 1)));
  branch2->push_back(BasicConv2d(Options(ch3x3red, ch3x3, 3).padding(1)));

  branch3->push_back(BasicConv2d(Options(in_channels, ch5x5red, 1)));
  branch3->push_back(BasicConv2d(Options(ch5x5red, ch5x5, 3).padding(1)));

  branch4->push_back(
      torch::nn::Functional(torch::max_pool2d, 3, 1, 1, 1, true));
  branch4->push_back(BasicConv2d(Options(in_channels, pool_proj, 1)));

  register_module("branch1", branch1);
  register_module("branch2", branch2);
  register_module("branch3", branch3);
  register_module("branch4", branch4);
}

torch::Tensor InceptionImpl::forward(torch::Tensor x) {
  auto b1 = branch1->forward(x);
  auto b2 = branch2->forward(x);
  auto b3 = branch3->forward(x);
  auto b4 = branch4->forward(x);

  return torch::cat({b1, b2, b3, b4}, 1);
}

InceptionAuxImpl::InceptionAuxImpl(int64_t in_channels, int64_t num_classes) {
  conv = BasicConv2d(Options(in_channels, 128, 1));
  fc1 = torch::nn::Linear(2048, 1024);
  fc2 = torch::nn::Linear(1024, num_classes);

  register_module("conv", conv);
  register_module("fc1", fc1);
  register_module("fc2", fc2);
}

torch::Tensor InceptionAuxImpl::forward(at::Tensor x) {
  // aux1: N x 512 x 14 x 14, aux2: N x 528 x 14 x 14
  x = torch::adaptive_avg_pool2d(x, {4, 4});
  // aux1: N x 512 x 4 x 4, aux2: N x 528 x 4 x 4
  x = conv->forward(x);
  // N x 128 x 4 x 4
  x = x.view({x.size(0), -1});
  // N x 2048
  x = fc1->forward(x).relu_();
  // N x 2048
  x = torch::dropout(x, 0.7, is_training());
  // N x 2048
  x = fc2->forward(x);
  // N x 1024

  return x;
}

} // namespace _googlenetimpl

GoogLeNetImpl::GoogLeNetImpl(
    int64_t num_classes,
    bool aux_logits,
    bool transform_input,
    bool init_weights) {
  this->aux_logits = aux_logits;
  this->transform_input = transform_input;

  conv1 = _googlenetimpl::BasicConv2d(Options(3, 64, 7).stride(2).padding(3));
  conv2 = _googlenetimpl::BasicConv2d(Options(64, 64, 1));
  conv3 = _googlenetimpl::BasicConv2d(Options(64, 192, 3).padding(1));

  inception3a = _googlenetimpl::Inception(192, 64, 96, 128, 16, 32, 32);
  inception3b = _googlenetimpl::Inception(256, 128, 128, 192, 32, 96, 64);

  inception4a = _googlenetimpl::Inception(480, 192, 96, 208, 16, 48, 64);
  inception4b = _googlenetimpl::Inception(512, 160, 112, 224, 24, 64, 64);
  inception4c = _googlenetimpl::Inception(512, 128, 128, 256, 24, 64, 64);
  inception4d = _googlenetimpl::Inception(512, 112, 144, 288, 32, 64, 64);
  inception4e = _googlenetimpl::Inception(528, 256, 160, 320, 32, 128, 128);

  inception5a = _googlenetimpl::Inception(832, 256, 160, 320, 32, 128, 128);
  inception5b = _googlenetimpl::Inception(832, 384, 192, 384, 48, 128, 128);

  if (aux_logits) {
    aux1 = _googlenetimpl::InceptionAux(512, num_classes);
    aux2 = _googlenetimpl::InceptionAux(528, num_classes);

    register_module("aux1", aux1);
    register_module("aux2", aux2);
  }

  dropout = torch::nn::Dropout(0.2);
  fc = torch::nn::Linear(1024, num_classes);

  register_module("conv1", conv1);
  register_module("conv2", conv2);
  register_module("conv3", conv3);

  register_module("inception3a", inception3a);
  register_module("inception3b", inception3b);

  register_module("inception4a", inception4a);
  register_module("inception4b", inception4b);
  register_module("inception4c", inception4c);
  register_module("inception4d", inception4d);
  register_module("inception4e", inception4e);

  register_module("inception5a", inception5a);
  register_module("inception5b", inception5b);

  register_module("dropout", dropout);
  register_module("fc", fc);

  if (init_weights)
    _initialize_weights();
}

void GoogLeNetImpl::_initialize_weights() {
  for (auto& module : modules(/*include_self=*/false)) {
    if (auto M = dynamic_cast<torch::nn::Conv2dImpl*>(module.get()))
      torch::nn::init::normal_(M->weight); // Note: used instead of truncated
                                           // normal initialization
    else if (auto M = dynamic_cast<torch::nn::LinearImpl*>(module.get()))
      torch::nn::init::normal_(M->weight); // Note: used instead of truncated
                                           // normal initialization
    else if (auto M = dynamic_cast<torch::nn::BatchNormImpl*>(module.get())) {
      torch::nn::init::ones_(M->weight);
      torch::nn::init::zeros_(M->bias);
    }
  }
}

GoogLeNetOutput GoogLeNetImpl::forward(torch::Tensor x) {
  if (transform_input) {
    auto x_ch0 = torch::unsqueeze(x.select(1, 0), 1) * (0.229 / 0.5) +
        (0.485 - 0.5) / 0.5;
    auto x_ch1 = torch::unsqueeze(x.select(1, 1), 1) * (0.224 / 0.5) +
        (0.456 - 0.5) / 0.5;
    auto x_ch2 = torch::unsqueeze(x.select(1, 2), 1) * (0.225 / 0.5) +
        (0.406 - 0.5) / 0.5;

    x = torch::cat({x_ch0, x_ch1, x_ch2}, 1);
  }

  // N x 3 x 224 x 224
  x = conv1->forward(x);
  // N x 64 x 112 x 112
  x = torch::max_pool2d(x, 3, 2, 0, 1, true);
  // N x 64 x 56 x 56
  x = conv2->forward(x);
  // N x 64 x 56 x 56
  x = conv3->forward(x);
  // N x 192 x 56 x 56
  x = torch::max_pool2d(x, 3, 2, 0, 1, true);

  // N x 192 x 28 x 28
  x = inception3a->forward(x);
  // N x 256 x 28 x 28
  x = inception3b->forward(x);
  // N x 480 x 28 x 28
  x = torch::max_pool2d(x, 3, 2, 0, 1, true);
  // N x 480 x 14 x 14
  x = inception4a->forward(x);
  // N x 512 x 14 x 14
  torch::Tensor aux1;
  if (is_training() && aux_logits)
    aux1 = this->aux1->forward(x);

  x = inception4b->forward(x);
  // N x 512 x 14 x 14
  x = inception4c->forward(x);
  // N x 512 x 14 x 14
  x = inception4d->forward(x);
  // N x 528 x 14 x 14
  torch::Tensor aux2;
  if (is_training() && aux_logits)
    aux2 = this->aux2->forward(x);

  x = inception4e(x);
  // N x 832 x 14 x 14
  x = torch::max_pool2d(x, 2, 2, 0, 1, true);
  // N x 832 x 7 x 7
  x = inception5a(x);
  // N x 832 x 7 x 7
  x = inception5b(x);
  // N x 1024 x 7 x 7

  x = torch::adaptive_avg_pool2d(x, {1, 1});
  // N x 1024 x 1 x 1
  x = x.view({x.size(0), -1});
  // N x 1024
  x = dropout->forward(x);
  x = fc->forward(x);
  // N x 1000(num_classes)

  return {x, aux1, aux2};
}

} // namespace models
} // namespace vision
